sklearn/preprocessing/_encoders.py

```diff
@@ -397,6 +496,190 @@ def _compute_drop_idx(self):
+                             "infrequent, try decreasing min_frequency")
+        return output if output.size > 0 else None
+
+    def _fit_infrequent_category_mapping(self, fit_results):
+        """Fit infrequent categories.
+
+        Defines the private attribute: `_default_to_infrequent_mappings`. For
+        feature `i`, `_default_to_infrequent_mappings[i]` defines the mapping
+        from the integer encoding returned by `super().transform()` into
+        infrequent categories. If `_default_to_infrequent_mappings[i]` is None,
+        there were no infrequent categories in the training set.
+
+        Parameters
+        ----------
+        fit_results : dict
+            return values from `super()._fit()`
+        """
+        if not self._infrequent_enabled:
+            return
+
+        n_samples = fit_results["n_samples"]
+        category_counts = fit_results["category_counts"]
+
+        self.infrequent_indices_ = [
+            self._identify_infrequent(category_count, n_samples, col_idx)
+            for col_idx, category_count in enumerate(category_counts)
+        ]
+
+        # compute mapping from default mapping to infrequent mapping
+        default_to_infrequent_mappings = []
+
+        for cats, infreq_idx in zip(self.categories_,
+                                    self.infrequent_indices_):
+            # no infrequent categories
+            if infreq_idx is None:
+                default_to_infrequent_mappings.append(None)
+                continue
+
+            n_cats = len(cats)
+            # infrequent indices exist
+            mapping = np.empty(n_cats, dtype=int)
+            n_infrequent_cats = infreq_idx.size
+
+            # infrequent categories are mapped to the last element.
+            n_frequent_cats = n_cats - n_infrequent_cats
+            mapping[infreq_idx] = n_frequent_cats
+
+            frequent_indices = np.setdiff1d(np.arange(n_cats), infreq_idx)
+            mapping[frequent_indices] = np.arange(n_frequent_cats)
+
+            default_to_infrequent_mappings.append(mapping)
+
+        self._default_to_infrequent_mappings = default_to_infrequent_mappings
+
+    def _map_to_infrequent_categories(self, X_int, X_mask):
+        """Map categories to infrequent categories. This modifies X_int
+        in-place.
+
+        Parameters
+        ----------
+        X_int: ndarray of shape (n_samples, n_features)
+            Integer encoded categories.
+        """
+        if not self._infrequent_enabled:
+            return
+
+        n_features = X_int.shape[1]
+        for col_idx in range(n_features):
+            infrequent_idx = self.infrequent_indices_[col_idx]
+            if infrequent_idx is None:
+                continue
+
+            X_int[~X_mask[:, col_idx], col_idx] = infrequent_idx[0]
+            if self.handle_unknown == 'auto':
+                # unknown values will be mapped to infrequent in the next for
+                # loop
+                X_mask[:, col_idx] = True
```

Comment:
I am confused by this. `X_mask[:, col_idx]` seems unused after this line within this method.